src/Microsoft.AspNet.Security.OAuthBearer/OAuthBearerAuthenticationHandler.cs

```diff
@@ -0,0 +1,195 @@
+                {
+                    _configuration = await Options.ConfigurationManager.GetConfigurationAsync(Context.RequestAborted);
+                }
+
+                var validationParameters = Options.TokenValidationParameters.Clone();
+                if (_configuration != null)
+                {
+                    if (validationParameters.ValidIssuer == null && !string.IsNullOrWhiteSpace(_configuration.Issuer))
+                    {
+                        validationParameters.ValidIssuer = _configuration.Issuer;
+                    }
+                    else
+                    {
+                        IEnumerable<string> issuers = new[] { _configuration.Issuer };
+                        validationParameters.ValidIssuers = (validationParameters.ValidIssuers == null ? issuers : validationParameters.ValidIssuers.Concat(issuers));
+                    }
+
+                    validationParameters.IssuerSigningKeys = (validationParameters.IssuerSigningKeys == null ? _configuration.SigningKeys : validationParameters.IssuerSigningKeys.Concat(_configuration.SigningKeys));
+                }
+
+                SecurityToken validatedToken;
+                foreach (var validator in Options.SecurityTokenValidators)
+                {
+                    if (validator.CanReadToken(token))
+                    {
+                        ClaimsPrincipal principal = validator.ValidateToken(token, validationParameters, out validatedToken);
+                        AuthenticationTicket ticket = new AuthenticationTicket(principal, new AuthenticationProperties(), Options.AuthenticationType);
+                        var securityTokenValidatedNotification = new SecurityTokenValidatedNotification<HttpContext, OAuthBearerAuthenticationOptions>(Context, Options)
+                        {
+                            ProtocolMessage = Context,
+                            AuthenticationTicket = ticket
+                        };
+
+                        if (securityTokenReceivedNotification.HandledResponse)
+                        {
+                            return securityTokenValidatedNotification.AuthenticationTicket;
+                        }
+
+                        if (securityTokenReceivedNotification.Skipped)
+                        {
+                            return null;
+                        }
+
+                        return ticket;
+                    }
+                }
+
+                throw new InvalidOperationException("No SecurityTokenValidator available for token: " + token ?? "null");
+            }
+            catch (Exception ex)
+            {
+                // We can't await inside a catch block, capture and handle outside.
+                authFailedEx = ExceptionDispatchInfo.Capture(ex);
+            }
+
+            if (authFailedEx != null)
+            {
+                _logger.WriteError("Exception occurred while processing message: '" + authFailedEx.ToString());
```

Comment:
There is a `WriteError` extension that can take in the `exception` object directly. https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Framework.Logging/LoggerExtensions.cs#L333